src/jit.rs

```diff
@@ -129,115 +129,99 @@ const REGISTER_MAP: [u8; 11] = [
+        *ptr = data as T;
+    }
+    jit.offset_in_text_section += size;
+    Ok(())
 }
 
 #[allow(dead_code)]
 #[inline]
-fn emit_debugger_trap(jit: &mut JitCompiler) {
-    emit1(jit, 0xcc);
+fn emit_debugger_trap<E: UserDefinedError>(jit: &mut JitCompiler) -> Result<(), EbpfError<E>> {
+    emit::<u8, E>(jit, 0xcc)
 }
 
 #[inline]
-fn emit_modrm(jit: &mut JitCompiler, modrm: u8, r: u8, m: u8) {
-    assert_eq!((modrm | 0xc0), 0xc0);
-    emit1(jit, (modrm & 0xc0) | ((r & 0b111) << 3) | (m & 0b111));
+fn emit_modrm<E: UserDefinedError>(jit: &mut JitCompiler, modrm: u8, r: u8, m: u8) -> Result<(), EbpfError<E>> {
+    debug_assert_eq!((modrm | 0xc0), 0xc0);
+    emit::<u8, E>(jit, (modrm & 0xc0) | ((r & 0b111) << 3) | (m & 0b111))
 }
 
 #[inline]
-fn emit_modrm_reg2reg(jit: &mut JitCompiler, r: u8, m: u8) {
-    emit_modrm(jit, 0xc0, r, m);
+fn emit_modrm_reg2reg<E: UserDefinedError>(jit: &mut JitCompiler, r: u8, m: u8) -> Result<(), EbpfError<E>> {
+    emit_modrm(jit, 0xc0, r, m)
 }
 
 #[inline]
-fn emit_sib(jit: &mut JitCompiler, scale: u8, index: u8, base: u8) {
-    assert_eq!((scale | 0xc0), 0xc0);
-    emit1(jit, (scale & 0xc0) | ((index & 0b111) << 3) | (base & 0b111));
+fn emit_sib<E: UserDefinedError>(jit: &mut JitCompiler, scale: u8, index: u8, base: u8) -> Result<(), EbpfError<E>> {
+    debug_assert_eq!((scale | 0xc0), 0xc0);
+    emit::<u8, E>(jit, (scale & 0xc0) | ((index & 0b111) << 3) | (base & 0b111))
 }
 
 #[inline]
-fn emit_modrm_and_displacement(jit: &mut JitCompiler, r: u8, m: u8, d: i32) {
+fn emit_modrm_and_displacement<E: UserDefinedError>(jit: &mut JitCompiler, r: u8, m: u8, d: i32) -> Result<(), EbpfError<E>> {
     if d == 0 && (m & 0b111) != RBP {
-        emit_modrm(jit, 0x00, r, m);
+        emit_modrm(jit, 0x00, r, m)?;
         if (m & 0b111) == RSP {
-            emit_sib(jit, 0, m, m);
+            emit_sib(jit, 0, m, m)?;
         }
     } else if d >= -128 && d <= 127 {
-        emit_modrm(jit, 0x40, r, m);
+        emit_modrm(jit, 0x40, r, m)?;
         if (m & 0b111) == RSP {
-            emit_sib(jit, 0, m, m);
+            emit_sib(jit, 0, m, m)?;
         }
-        emit1(jit, d as u8);
+        emit::<u8, E>(jit, d as u8)?;
     } else {
-        emit_modrm(jit, 0x80, r, m);
+        emit_modrm(jit, 0x80, r, m)?;
         if (m & 0b111) == RSP {
-            emit_sib(jit, 0, m, m);
+            emit_sib(jit, 0, m, m)?;
         }
-        emit4(jit, d as u32);
+        emit::<u32, E>(jit, d as u32)?;
     }
+    Ok(())
 }
 
 #[inline]
-fn emit_rex(jit: &mut JitCompiler, w: u8, r: u8, x: u8, b: u8) {
-    assert_eq!((w | 1), 1);
-    assert_eq!((r | 1), 1);
-    assert_eq!((x | 1), 1);
-    assert_eq!((b | 1), 1);
-    emit1(jit, 0x40 | (w << 3) | (r << 2) | (x << 1) | b);
+fn emit_rex<E: UserDefinedError>(jit: &mut JitCompiler, w: u8, r: u8, x: u8, b: u8) -> Result<(), EbpfError<E>> {
+    debug_assert_eq!((w | 1), 1);
```

Comment:
These could be encountered in a user deployed program?  If so, how would they be handled if not caught here?